### ISSUE TITLE
[autoperf-iter7] attn_proj_out checkpoint offload produces NaN at step 1 in production training

### ISSUE THREAD

**Issue by ultrons** [ultrons/jax-gpt#2]
**Discovered by**: autoperf agent, iter-7 (2026-05-08), commit `8245f5d` → reverted `bbfe974`.
**Cross-ref**: `autoperf/iter_log.md` § iter 7, `autoperf/v7x_KNOWLEDGE.md` §3 "attn_proj_out offload broken at production scale".
**Status**: workaround in place (don't add `attn_proj_out` to offload list). Real fix needed for iter-N to claim 4,216 ms/step of attention-recompute headroom.

## Repro

One-line change at `jax_gpt/models/dsv3/model.py:3054`:

```diff
 _ckpt_policy = jax.checkpoint_policies.save_and_offload_only_these_names(
     names_which_can_be_saved=(),
-    names_which_can_be_offloaded=("moe_layer_input",),
+    names_which_can_be_offloaded=("moe_layer_input", "attn_proj_out"),
     offload_src="device",
     offload_dst="pinned_host",
 )
```

The marker `out = ck(out, "attn_proj_out")` already exists at `model.py:560` (CP path) and `:636` (non-CP path). The change just adds it to the offload list.

**Cluster command** (image `cde-64df9c2` was the iter-7 build of this change):
```
cde build
cde run --tag dsv3train-i7 --context gke_cloud-tpu-multipod-dev_us-central1_bodaborg-super-rbq --profile \
  --set config=full --set moe_backend=gmm_ag --set attn_backend=splash --set optimizer=sgd \
  --set fsdp=128 --set ep=4 --set tp=1 --set gbs=4096 --set grad_accum=1 \
  --set moe_n_chunks=2 --set steps=5 --set grad_clip=1.0 --set dtype=bfloat16 \
  --set gradient_checkpoint=true --set no_cp=true --set moe_use_gmm_v2=true
```

## Symptom

All 5 steps report `loss: nan (lm=nan, aux=nan)` from step 1 onward. The cluster JobSet reports `Finished=True (Succeeded)` because the Python script exits 0; only the per-step log lines reveal the failure.

Step 4 measurements:
```
completed step: 4, seconds: 34.119, TFLOP/s/chip: 717.1, MFU: 31.1%, TPS/chip: 1920.8, loss: nan (lm=nan, aux=nan)
```

The **+2% TPS metric is illusory** — with NaN gradients propagating, every multiply produces NaN, every reduction produces NaN. These are bit-pattern-only operations with zero numerical work, uniformly fast. Don't be fooled by the speedup.

## Pre-flight verification that PASSED but didn't catch this

1. **AOT compile gate** with production `LIBTPU_INIT_ARGS=--xla_tpu_scoped_vmem_limit_kib=65536` mirrored — passed.
2. **`jax.make_jaxpr` verification** on a toy fn:
   - `dot_general` count in grad-jaxpr drops from 10 → 9 (Δ=-1) with the change vs without
   - `attn_proj_out` appears as a saved residual in the bwd jaxpr (specifically as `from_residual` / offload), not as a recomputed op
   - Confirms the policy is doing the right thing at the IR level

3. **Policy parsing** — `save_and_offload_only_these_names(names_which_can_be_offloaded=("moe_layer_input", "attn_proj_out"), ...)` constructs cleanly.

The NaN is **runtime-only**, manifests only on the actual TPU cluster, not at AOT/jaxpr time.

## Why moe_layer_input works but attn_proj_out doesn't (hypothesis)

`moe_layer_input` is named at the **start** of each MoE layer (`model.py:2923`). Between when forward computes it and bwd needs it, the entire layer's compute happens (~5,000 ms attention + MoE). The async DMA to host has ample slack to complete.

`attn_proj_out` is named **mid-layer**, after attention finishes. Between its computation and bwd's read, only the MoE forward runs (~1,800 ms). The DMA may be still in flight when the bwd starts reading.

## Hypotheses on root cause (in priority order)

1. **Async `pinned_host` DMA race**: the bwd's host-DMA read starts before the fwd's host-DMA write has completed. Bwd reads uninitialized memory → NaN propagates. Most likely given the slack analysis above.
2. **bf16↔fp32 conversion drift on offload roundtrip**: layout/precision changes during host transfer combine with attention's softmax/sqrt to produce NaN. Less likely but possible.
3. **Sharded-layout mismatch when offloaded value is restored**: the activation has multi-axis sharding `(B_l, S, D)` bf16 that may not round-trip through `pinned_host` cleanly.

## Related markers that may have the same issue

The v315 author marked several activations with `checkpoint_name` but only wired `moe_layer_input` into the offload policy. Their comment at `model.py:3047-3051` reads:
> "Only large activations (moe_layer_input, attn_proj_out) have favorable DUS:save ratio"

Now confirmed: `attn_proj_out` is broken (this issue). The other markers — `q_a` (`model.py:568`), `kv_a` (around `:580`), `shared_hidden` (`model.py:2676`) — were tested and rejected by v315 for "small DUS:save ratio" but the NaN failure mode might also apply to them. Don't enable any of these markers in the offload policy without verifying numerical correctness on cluster.

## Definition of done

1. `attn_proj_out` can be added to `names_which_can_be_offloaded` and training produces finite loss matching iter-2b baseline (415.46 ± bf16 tolerance) at step 1.
2. Steady-state step time delta vs iter-2b is positive (offload short-circuit > DUS overhead).
3. Documented root cause (async DMA timing? layout? conversion?) and fix in a way that informs whether `q_a` / `kv_a` / `shared_hidden` can also be safely added.

## Workaround (in place)

Don't add `attn_proj_out` to `names_which_can_be_offloaded`. iter-2b state (only `moe_layer_input` offloaded) remains the verified production baseline at 1882 TPS/chip @ 30.5% MFU. The 4,216 ms/step of attention-recompute headroom (per `research/dsv3/iter6_checkpoint_bisection.md`) is unaddressable from the offload-policy side until this bug is fixed.

## Cluster cost of the failed iter

~15 min total (admission + compile + 5 NaN steps + 600s sleep buffer). Cleanly recoverable; nothing destabilized.

## Priority

P2 — autoperf-blocking-but-workaroundable. Production training works fine without this lever; iter-N can pursue other levers (attention-only-checkpoint refactor, scatter fusion, etc.) while this is fixed.
